Introducing `/quiz` This is a matching quiz to see what acm team you fit in

AlanCortez1337 opened the pull request:
I want to make a matching app that helps choose possible events to go to for new people. While showing off resources for that and ways they can join/help the acm board.

AlanCortez1337 committed:
👉👈🙄

AlanCortez1337 committed:
first iteration of results page

AlanCortez1337 committed:
finished first iteration of styles and interaction elements for the quiz component

EthanThatOneKid commented on src/lib/components/icons/leftArrow.svelte:
```diff
@@ -0,0 +1,14 @@
+<script lang="ts">
```
Nit: Follow file naming convention. For example, this file should be named `left-arrow.svelte`.

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/icons/rightArrow.svelte:
```diff
@@ -0,0 +1,14 @@
+<script lang="ts">
```
Nit: Follow file naming convention.

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/moreInfo.svelte:
```diff
@@ -0,0 +1,6 @@
+<script lang="ts">
```
Nit: Follow file naming convention.

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,233 @@
+<script lang="ts">
+  import type { QuizData } from '$lib/constants/quiz';
+  import LeftArrow from '../icons/leftArrow.svelte';
+  import RightArrow from '../icons/rightArrow.svelte';
```
To import common icons, you can use `$lib`.
```suggestion
  import LeftArrow from '$lib/components/icons/left-arrow.svelte';
  import RightArrow from '$lib/components/icons/right-arrow.svelte';
```

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/constants/quiz.ts:
```diff
@@ -0,0 +1,98 @@
+  // Where the choice is an index number for use in quiz.svelte
+  // ai => 0, dev => 1, design => 2, algo => 3, N/A => 4
+  match: string;
+}
```
We can use an enum to make sure the quiz is being accurately calculated.

```suggestion
export enum TeamMatch {
  TEAMLESS,
  ALGO,
  DEV,
  DESIGN,
  AI,
}

interface Choice {
  content: string;
  // Where color is a valid HTML color
  color: string;
  match: TeamMatch;
}
```

AlanCortez1337 replied:
Done, reordered enums and gave them unique values so it properly works.

EthanThatOneKid replied:
LGTM. But...

Note: The reason why I ordered the enum with `TEAMLESS` first is because when left alone, the enum values are 0, 1, 2, 3, and so on. Usually, you want the _missing_ or _unsupported_-like value of your enum to be at the top, as I suggested, because the first value is 0. This is therefore the most consistent placement in the enum. 0 is also used as an initialization value in most languages.

EthanThatOneKid commented on src/lib/constants/quiz.ts:
```diff
@@ -0,0 +1,98 @@
+        {
+          content: 'ai',
+          color: '#21D19F',
+          match: 'aiChoice',
```
Implement enum `TeamMatch` in `/src/lib/constants/quiz.ts`.
```suggestion
          match: TeamMatch.AI,
```

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,233 @@
+  export let data: QuizData;
+
+  $: index = 0;
+  let response: string[] = [];
```
We are currently using variables `tallyChoices` and `response` to keep track of the quiz score. Since this is a quiz, the quiz-taker is creating a list of responses. To keep the calculations simple, let's redeclare `response`.
```suggestion
  let responses: TeamMatch[] = [];
```
Note: Make sure `TeamMatch` is added to your import of `QuizData`.

jaasonw replied:
maybe also avoid team quiz specific naming so this can be reused for other quizzes. say for instance algo wanted to have a review quiz at the end of the whatever?

EthanThatOneKid replied:
Good thinking. We have a file of constants in [`/src/lib/constants/acm-paths.ts `](https://github.com/EthanThatOneKid/acmcsuf.com/blob/461c99785d165af72fd5125921ff2f6d252168a8/src/lib/constants/acm-paths.ts), but “paths” have been renamed to “teams”.

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,233 @@
+
+  $: index = 0;
+  let response: string[] = [];
+  let noMoreQuestions: boolean = false;
```
Nit: This variable is unused.

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,233 @@
+  $: index = 0;
+  let response: string[] = [];
+  let noMoreQuestions: boolean = false;
+  let answeredEveryQuestion: boolean = false;
```
Once the user has answered every question, a third button should appear that leads to the results page.
We might be able to use a reactive statement to omit the checks in your `recordAnswer` function.
```suggestion
  $: answeredAllQuestions = responses.length === data.questions.length && !responses.includes(undefined);
```

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,233 @@
+      {/if}
+      <section class="answers">
+        {#each data.questions[index].choices as choice, j (choice.content)}
+          <button on:click={() => recordAnswer(j)} style={`--color: ${choice.color}`}>
```
We could pass `match` to the `recordAnswer` function directly.
```ts
  function recordAnswer(match: TeamMatch) {
```
```suggestion
        {#each data.questions[index].choices as choice (choice.content)}
          <button on:click={() => recordAnswer(choice.match)} style={`--color: ${choice.color}`}>
```

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,233 @@
+    algoChoice: 0,
+    designChoice: 0,
+    devChoice: 0,
+  };
```
We might be able to use a complicated reactive statement to calculate the quiz result.
```suggestion
  $: talliedResponses = responses.reduce((tallies, match, i) => {
    if (tallies[match]) tallies[match]++;
    else tallies[match] = 1;
    return tallies;
  }, {});
```

AlanCortez1337 replied:
Done, removed the i parameter since Svelte was yelling at me that it was not being used when I tried npm run all.

EthanThatOneKid replied:
Good catch; not sure how that got in there 😅

EthanThatOneKid reviewed (commented):
I know the PR is not finished yet, but in the mean time I took a look and have some suggestions @AlanCortez1337 to follow up on.

AlanCortez1337 committed:
refractored logic to fit recommendations, began styling results page, updated styles a bit for the quiz component

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,249 @@
+      <h2>{data.questions[index].prompt}</h2>
+      {#if response[index]}
+        <p>answered</p>
+      {/if}
```
Perhaps, allow the user to update their responses until they hit the submit button.

AlanCortez1337 replied:
They can update their responses already, this is merely a quick fix to let them know that they already answered the question just in case they forgot. I can see why this is misleading so don't resolve this issue I will come up with a better UI solution.

AlanCortez1337 replied:
Done, now there are unique styles for selected answers

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,249 @@
+  export let data: QuizData;
+
+  $: index = 0;
+  let response: TeamMatch[] = [];
```
Since the variable `response` holds a list of responses, we could name it something plural e.g. `responses` or `choices`.

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/more-info.svelte:
```diff
@@ -0,0 +1,6 @@
+<script lang="ts">
+  export let teamPage = 'help';
```
This component will show more info given a team match.
```suggestion
  import type { TeamMatch } from '$lib/constants/quiz';

  export let teamMatch: TeamMatch = TeamMatch.TEAMLESS;
```

AlanCortez1337 replied:
Done, unnecessary after final implementation

EthanThatOneKid commented on src/lib/components/quiz/more-info.svelte:
```diff
@@ -0,0 +1,6 @@
+</script>
+
+<h1>This is {teamPage}</h1>
+<h3>We do Stuff!</h3>
```
TODO: Display different content depending on the matching team variable.

AlanCortez1337 replied:
Done

EthanThatOneKid replied:
LGTM!
<https://github.com/EthanThatOneKid/acmcsuf.com/blob/203c112c377efbaa99790779d85161621b6cbdf5/src/lib/components/quiz/more-info.svelte#L7-L41>

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,249 @@
+
+  export let data: QuizData;
+
+  $: index = 0;
```
Since we want to initially setting the variable `index` to `0`, we do not need to set this variable with the _reactive_ syntax.
```suggestion
  let index = 0;
```

AlanCortez1337 replied:
done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,249 @@
+    return tallies;
+  }, {});
+  let teams = ['Algo', 'Dev', 'Design', 'AI'];
+  let match: string;
```
We can calculate the resulting team match reactively, using `talliedResponses` as a dependency.
```suggestion
  $: match = Object.entries(talliedResponses).sort(([,a], [,b]) => b - a).shift().shift();
```

AlanCortez1337 replied:
done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,249 @@
+    // if answered all questions and no more questions then submitResponses
+    if (!answeredAllQuestions) {
+      goRight();
+    }
```
The `goRight` function might contain sufficient checks before going right.
```suggestion
    goRight();
```

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,249 @@
+    // save to localstorage
+    // show results and hide the question
+    showResults = true;
+  }
```
The function that is called when submitting the quiz should simply show the results screen. Calculations are made reactively.
```suggestion
  function onSubmitResponses() {
    // show results and hide the question
    showResults = true;
  }
```

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,249 @@
+  $: index = 0;
+  let response: TeamMatch[] = [];
+  $: answeredAllQuestions =
+    response.length === data.questions.length && !response.includes(undefined);
```
We can build a reactive statement that automatically stores the user's responses in [local storage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage).
```suggestion
  let answeredAllQuestions = false;
  $: {
    answeredAllQuestions =
      responses.length === data.questions.length && !responses.includes(undefined);
    // Updates local storage whenever `responses` changes.
    quizStorage.setResponses(responses);
  }
```

AlanCortez1337 replied:
done

EthanThatOneKid reviewed (changes requested):
I left some nits this morning. I am still looking at the changes, but this is amazing progress!

EthanThatOneKid committed:
Refactored $lib/quiz

- Removed $lib/constants/quiz
- Isolated quiz data in separate file
- Created quiz responses storage service
- Isolated types in separate file
- Updated changed imports

Note: All quiz-related code shall be exported from the index.ts file e.g. from $lib/quiz.

AlanCortez1337 committed:
make first iteration of progress bar and fixed some comments

AlanCortez1337 committed:
addressed some comments, finished second iteration of results page, made progress-bar component

AlanCortez1337 committed:
addressed some comments, finished second iteration of results page, made progress-bar component

AlanCortez1337 committed:
implemented local storage logic, refractored how to store team data, refined styles

AlanCortez1337 committed:
finished first iteration of more-info.svelte, changed some styles in the results section, added images for team workshops

EthanThatOneKid commented on src/lib/constants/acm-paths.ts:
```diff
@@ -3,41 +3,75 @@ export interface AcmPath {
+  blurp: string;
+  blurpRecommend: string;
+  recommendations: string[];
+  workshopPictures: string[];
```
Neat idea to add more team-specific constants. I also reworded slightly.
```suggestion
  description: string;
  recommendationsBlurb: string;
  recommendations: Array<{ title: string; link: string; }>;
  workshopPictures: string[];
```

AlanCortez1337 replied:
Done

EthanThatOneKid replied:
The word is "blurb", I believe. But `blurp` is fine if you insist.

EthanThatOneKid commented on static/quiz/frank-js-emote.svg:
```diff
@@ -0,0 +1,47 @@
+<svg width="400" height="342" viewBox="0 0 400 342" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Please make sure that the name of the files in `/static` are completely lowercased.

AlanCortez1337 replied:
Done

EthanThatOneKid replied:
Git is a little tricky with renaming file names, so we can resolve the stragglers in a future change.

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,357 @@
+
+  onMount(() => {
+    // hypothetically change this to QuizStorage.init()
+    quizStorage = new QuizStorage();
```
Load the visitor's previous answers if they exist in local storage.
```suggestion
    quizStorage = new QuizStorage();
    responses = quizStorage.getResponses();
```

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,357 @@
+  import LeftArrow from '$lib/components/icons/left-arrow.svelte';
+  import RightArrow from '$lib/components/icons/right-arrow.svelte';
+  import MoreInfo from './more-info.svelte';
+  import AcmPaths from '../index/acm-paths.svelte';
```
Make sure to run `npm run all` before committing your changes. The website check caught an error.
> `'AcmPaths' is defined but never used`

![image](https://user-images.githubusercontent.com/31261035/188996159-109e9be0-40ca-4956-9c7e-79246be561d5.png)

AlanCortez1337 replied:
Done

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,357 @@
+  $: match = (Object.entries(talliedResponses)
+    .sort(([, a], [, b]) => b - a)
+    ?.shift()
+    ?.shift() ?? TeamMatch.TEAMLESS) as TeamMatch;
```
Nit: Since these declarations take up more than one line each, I'd separate them.
```suggestion
  $: talliedResponses = responses.reduce((tallies, match) => {
    if (tallies[match]) tallies[match]++;
    else tallies[match] = 1;
    return tallies;
  }, {} as Record<TeamMatch, number>);

  $: match = (Object.entries(talliedResponses)
    .sort(([, a], [, b]) => b - a)
    ?.shift()
    ?.shift() ?? TeamMatch.TEAMLESS) as TeamMatch;
```

AlanCortez1337 replied:
Done

EthanThatOneKid reviewed (changes requested):
Insane progress. I left some more nits for @AlanCortez1337.

AlanCortez1337 committed:
fixed styles, fixed some comments, and overall finished the final draft

AlanCortez1337 committed:
final draft, with npm run all

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,418 @@
+      >
+      <button
+        on:click={goRight}
+        disabled={index === data.questions.length - 1}
```
I think the "go right" button should be disabled until the user makes their choice.

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,418 @@
+      <button
+        on:click={goLeft}
+        disabled={index === 0}
+        class={`${index === 0 && 'disable-arrow'} arrow`}><LeftArrow /></button
```
I think this button element can contain the class `false`. To fix that, we can use a Svelte shorthand to conditionally assign a CSS class to an element.
```suggestion
        class:disable-arrow={index === 0}
        class="arrow"><LeftArrow /></button
```

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,418 @@
+      <button
+        on:click={goRight}
+        disabled={index === data.questions.length - 1}
+        class={`${index === data.questions.length - 1 && 'disable-arrow'} arrow`}
```
I think this button element can contain the class `false`. To fix that, we can use a Svelte shorthand to conditionally assign a CSS class to an element.
```suggestion
        class:disable-arrow={/* your conditional expression */}
        class="arrow"
```

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,418 @@
+    <button
+      on:click={submitResponses}
+      disabled={!answeredAllQuestions}
+      class={`${!answeredAllQuestions && 'disable-submitBTN'} submitBTN`}>Submit</button
```
I think this button element can contain the class `false`. To fix that, we can use a Svelte shorthand to conditionally assign a CSS class to an element.
```suggestion
      class:disable-submitBTN={!answeredAllQuestions}
      class="submitBTN">Submit</button
```

EthanThatOneKid commented on src/lib/components/quiz/quiz.svelte:
```diff
@@ -0,0 +1,418 @@
+    <!-- DISPLAY ADDIONTAL TEAM INFORMATION -->
+  {:else if showMoreInfo}
+    <MoreInfo teamMatch={showTeam} />
+    <button on:click={goBackToResults} class={`arrow return-to-results`}
```
The class attribute accepts double quotes when statically assigning.
```suggestion
    <button on:click={goBackToResults} class="arrow return-to-results"
```

EthanThatOneKid commented on src/lib/constants/acm-paths.ts:
```diff
@@ -3,41 +3,96 @@ export interface AcmPath {
   slug: string;
   picture: string;
   color: string;
+  blurp: string;
+  blurpRecommend: string;
+  recommendations: Array<{ title: string; link: string }>;
+  workshopPictures: string[];
```
Unfortunately, a side-effect of not creating your own teams object is that it may affect other code in the repository. By updating these constant variables, we have broken a test such that the expected values must be updated in order to pass the test. 

Similar to #475, I want you to generate 10 events from the existing `./src/routes/events/_testdata/events.ics` into ` ./src/routes/events/_testdata/events.ts` with the following methodology:

1. Uncomment the following code in `./src/lib/ical/parse.test.ts`

https://github.com/EthanThatOneKid/acmcsuf.com/blob/5745f3efd890b9884ca9c4b7b0a145d5e941e31c/src/lib/ical/parse.test.ts#L17-L20

2. Run `npm run all`
3. Recomment the code that was uncommented in step 1
4. Commit your changes

---
Related:
- Failed check: <https://github.com/EthanThatOneKid/acmcsuf.com/runs/8276329862?check_suite_focus=true#step:8:69>
- Recent occurrence of similar problem: <https://github.com/EthanThatOneKid/acmcsuf.com/pull/523#pullrequestreview-1092716870>
- Original solution: #475

EthanThatOneKid reviewed (changes requested):
Hopefully I am not forgetting anything 🤔! Pop off, @AlanCortez1337!

AlanCortez1337 committed:
finished final comments and followed npm run all fix

AlanCortez1337 committed:
PUSH IT TO PROD

AlanCortez1337 committed:
Merge branch 'main' into fix-527

EthanThatOneKid commented on src/lib/components/quiz/more-info.svelte:
```diff
@@ -0,0 +1,156 @@
+  export let workshopPictures: string[] = [];
+</script>
+
+<section class="container" style={`--team-color: ${color}`}>
```
I JUST remembered the Svelte shorthand I was thinking of when originally reviewing this code.
```html
<section class="container" style:--team-color={color}>
```
